### PR TITLE
uri_parser: constify result

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

`uri_parser_process()` takes a `const char *` as input parameter but the result is a struct of `char *` to the original string.

This may lead a user to modifying the strings in `uri_parser_result_t` which will cause a crash if the original string resides in read-only memory (and violates the no-modifications promise of the const parameter in `uri_parser_process()`).

To fix this, make the resulting strings `const` as well, so nobody dares to touch them in a writing way.

### Testing procedure

    make -C tests/unittests tests-uri_parser term

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/pull/16695#discussion_r681820654

### CODE REVIEW

**Review by miri64 — APPROVED** [review]
ACK. The parser is aiming to be non-destructive and since the input is `const`, so should the output.

**Comment by miri64** [comment]
> ### Testing procedure
> 
> ```
> make -C tests/unittests tests-uri_parser all term
> ```

So build only the `uri_parser` tests, then build all and then run the tests without automation :raised_eyebrow:?

**Comment by miri64** [comment]
Sadly, some of the users of this API apparently also don't use `const` :-/.

File: pkg/wakaama/contrib/lwm2m_client_connection.c [hunk 1]
```diff
@@ -220,10 +220,10 @@ static int _connection_send(lwm2m_client_connection_t *conn, uint8_t *buffer,
     return 0;
 }
 
-static netif_t *_get_interface(char *host)
+static netif_t *_get_interface(const char *host)
 {
     netif_t *netif = NULL;
-    char *iface = ipv6_addr_split_iface(host);
```
[the remark this answers is not in the document]
Comment (miri64) [reply pkg/wakaama/contrib/lwm2m_client_connection.c:226]: @benpicco see https://github.com/RIOT-OS/RIOT/pull/16709 for how to get the `netif` from the `zoneid` and `zoneid_len` parts of a `uri_parser_result_t`.

[the remark this answers is not in the document]
Comment (leandrolanzieri) [reply pkg/wakaama/contrib/lwm2m_client_connection.c:226]: For this particular case: https://github.com/RIOT-OS/RIOT/pull/16233/files#diff-f675597250e305b7990d9b3b02efd61e82e5b9ca09a6e2a57c04035023cafe49R242-R267

**Comment by miri64** [comment]
This PR needs #16709 anyways now, so postpone build, so #16709 can be merged first.

**Comment by benpicco** [comment]
Not sure how this is related to #16709

**Comment by miri64** [comment]
> Not sure how this is related to #16709

So you can use the new API here. See https://github.com/RIOT-OS/RIOT/pull/16707#discussion_r682701650.

File: pkg/wakaama/contrib/lwm2m_client_connection.c [hunk 2]
```diff
@@ -220,10 +220,10 @@ static int _connection_send(lwm2m_client_connection_t *conn, uint8_t *buffer,
...
+    const char *iface = strchr(host, '%') + 1;
```
Comment (miri64) [inline pkg/wakaama/contrib/lwm2m_client_connection.c:226]: Please use the `zoneid` from the `uri_parser_result_t` here.

**Review by miri64 — CHANGES_REQUESTED** [review]

File: pkg/wakaama/contrib/lwm2m_client_connection.c (hunk 2 above)
Comment (benpicco) [reply pkg/wakaama/contrib/lwm2m_client_connection.c:226]: I still don't see how this requires the new API - wouldn't this just be

```patch
--- a/pkg/wakaama/contrib/lwm2m_client_connection.c
+++ b/pkg/wakaama/contrib/lwm2m_client_connection.c
@@ -220,10 +220,9 @@ static int _connection_send(lwm2m_client_connection_t *conn, uint8_t *buffer,
     return 0;
 }
 
-static netif_t *_get_interface(const char *host)
+static netif_t *_get_interface(const char *iface)
 {
     netif_t *netif = NULL;
-    const char *iface = strchr(host, '%') + 1;
 
     if (iface == NULL) {
         /* get the number of net interfaces */
@@ -330,7 +329,7 @@ static lwm2m_client_connection_t *_connection_create(uint16_t sec_obj_inst_id,
      * if not, check the number of interfaces and default to the first if there
      * is only one defined. */
     if (ipv6_addr_is_link_local((ipv6_addr_t *)&conn->remote.addr.ipv6)) {
-        netif_t *netif = _get_interface(parsed_uri.host);
+        netif_t *netif = _get_interface(parsed_uri.zoneid);
         if (netif == NULL) {
             goto free_out;
         }
```

Comment (miri64) [reply pkg/wakaama/contrib/lwm2m_client_connection.c]: I see you found out, why the new API is required ;-).

**Review by miri64 — APPROVED** [review]
ACK

**Pull request merged**